### PR TITLE
py-tatsu: new package

### PR DESCRIPTION
**Pull request by healther**

### CODE REVIEW

**Comment by healther** [comment]
Depends on #13565 #13564 #13562

**Comment by healther** [comment]
Travis fail seems to be unrelated?

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    version('4.4.0', sha256='80713413473a009f2081148d0f494884cabaf9d6866b71f2a68a92b6442f343d')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-colorama@0.4:', type=('build', 'run'))
+    depends_on('py-dataclasses@0.6:', type=('build', 'run'), when='^python@3.6.0:3.6.99')
+    depends_on('py-mypy@0.641:', type=('build', 'run'))
+    depends_on('py-pyyaml@3.12:', type=('build', 'run'))
+    depends_on('py-regex@2018.8:', type=('build', 'run'))
+    depends_on('py-typing@3.6:', type=('build', 'run'), when='^python@:3.4.99')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]: Where did you find these requirements? They aren't in the `setup.py`.

Comment (healther) [reply var/spack/repos/builtin/packages/py-tatsu/package.py]: https://github.com/neogeny/TatSu/blob/master/requirements.txt

Comment (healther) [reply var/spack/repos/builtin/packages/py-tatsu/package.py]: I'm now really unsure on which we should follow... `requirements.txt` seems to be what `pip` is looking for, but it seems to be used in addition to what is state in `setup.py` and/or `setup.cfg`...

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-tatsu/package.py]: Only trust `setup.py` and `setup.cfg` in the current release tarball, never trust `requirements.txt` on master/develop.

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PyTatsu(PythonPackage):
+    """竜 TatSu (the successor to Grako) is a tool that takes grammars in
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]: Same thing here, Python 2 sucks for unicode, drop the kanji.

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    a variation of EBNF as input, and outputs memoizing (Packrat) PEG
+    parsers in Python."""
+
+    homepage = "https://github.com/neogeny/tatsu"
+    url      = "https://pypi.io/packages/source/T/TatSu/TatSu-4.4.0.zip"
+
+    version('4.4.0', sha256='80713413473a009f2081148d0f494884cabaf9d6866b71f2a68a92b6442f343d')
+
+    depends_on('python@3.6:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]: Remove this, Python 2 is still supported

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    depends_on('py-setuptools', type='build')
+    # part of the standard lib in python@3.7.0
+    depends_on('py-dataclasses@0.6:', type=('build', 'run'), when='^python@3.6.0:3.6.99')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]: I don't see this in the `setup.py`, remove it

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    depends_on('py-mypy@0.641:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]:
```suggestion
    depends_on('py-pytest-mypy', type='test')
```

File: var/spack/repos/builtin/packages/py-tatsu/package.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    # optional dependency, otherwise falls back to standard implementation
+    depends_on('py-regex@2018.8:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tatsu/package.py]: I would add a `+future_regex` variant for this and make it optional. Also, drop the version requirement.

Comment (healther) [reply var/spack/repos/builtin/packages/py-tatsu/package.py]: Done, but I'm not sure whether this is a good policy, I think in general we should prefer keeping build receipts simple over small. `spack` is good at building things, and it's not as if an additional python library is a significant time consumer.

**Comment by healther** [comment]
The additional requirements will come in for newer versions (ie they are in master, but not on any release yet), so I kept them around with a constraint to newer versions.

File: var/spack/repos/builtin/packages/py-tatsu/package.py (hunk 6 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-tatsu/package.py]: Depends on how many packages are optional and which they are. Some packages like numpy/pandas take quite a while to build. For all packages (not just Python) I like to make all optional dependencies truly optional.

**Pull request merged**